contracts/upgrader/src/contract.rs

```diff
@@ -0,0 +1,99 @@
+use crate::error::ContractError;
+use soroban_sdk::{contract, contractimpl, Address, BytesN, Env, String, Symbol, Val};
+
+#[contract]
+pub struct Upgrader;
+
+#[contractimpl]
+impl Upgrader {
+    pub fn __constructor(_env: Env) {}
+    pub fn upgrade(
+        env: Env,
+        contract_address: Address,
+        new_version: String,
+        new_wasm_hash: BytesN<32>,
+        migration_data: soroban_sdk::Vec<Val>,
+    ) -> Result<(), ContractError> {
+        assert_new_version_is_different(&env, &contract_address, &new_version)?;
+
+        env.invoke_contract::<()>(
+            &contract_address,
+            &upgrade_fn(&env),
+            soroban_sdk::vec![&env, new_wasm_hash.into()],
+        );
+        env.invoke_contract::<()>(&contract_address, &migrate_fn(&env), migration_data);
+
+        assert_new_version_matches_expected(&env, &contract_address, &new_version)
+    }
+}
+
+fn migrate_fn(env: &Env) -> Symbol {
+    Symbol::new(&env, "migrate")
+}
+
+fn upgrade_fn(env: &Env) -> Symbol {
+    Symbol::new(&env, "upgrade")
+}
+
+fn version_fn(env: &Env) -> Symbol {
+    Symbol::new(&env, "version")
+}
+
+fn assert_new_version_is_different(
+    env: &Env,
+    contract_address: &Address,
+    new_version: &String,
+) -> Result<(), ContractError> {
+    let cur_version: String = env.invoke_contract(
```

Comment:
perhaps define a function to get `current_version` of a contract and call it, as there is duplicated logic.
```
fn get_current_version(
    env: &Env,
    contract_address: &Address,
) -> Result<String, ContractError> {
    env.invoke_contract(
        contract_address,
        &version_fn(env),
        soroban_sdk::vec![env],
    ).map_err(|_| ContractError::InvocationError)
}

let cur_version = get_current_version(env, contract_address)?;
```